topology/src/test/java/io/camunda/zeebe/topology/changes/PartitionJoinApplierTest.java

```diff
@@ -0,0 +1,131 @@
+import io.camunda.zeebe.topology.state.PartitionState.State;
+import java.util.Map;
+import java.util.Set;
+import org.junit.jupiter.api.Test;
+
+final class PartitionJoinApplierTest {
+  private final PartitionTopologyChangeExecutor partitionTopologyChangeExecutor =
+      mock(PartitionTopologyChangeExecutor.class);
+  private final MemberId localMemberId = MemberId.from("1");
+  final ClusterTopology initialClusterTopology =
+      ClusterTopology.init().addMember(localMemberId, MemberState.initializeAsActive(Map.of()));
+  final PartitionJoinApplier partitionJoinApplier =
+      new PartitionJoinApplier(1, 1, localMemberId, partitionTopologyChangeExecutor);
+
+  @Test
+  void shouldRejectJoinIfPartitionIsAlreadyJoined() {
+    // given
+    final ClusterTopology topologyWithPartitionJoined =
+        initialClusterTopology.updateMember(
+            localMemberId, m -> m.addPartition(1, PartitionState.active(1)));
+
+    // when - then
+    assertThat(partitionJoinApplier.init(topologyWithPartitionJoined))
+        .isLeft()
+        .left()
+        .isInstanceOf(IllegalStateException.class);
+
+    // then
+  }
+
+  @Test
+  void shouldRejectJoinIfPartitionIsAlreadyLeaving() {
+    // given
+    final ClusterTopology topologyWithPartitionLeaving =
+        initialClusterTopology.updateMember(
+            localMemberId, m -> m.addPartition(1, PartitionState.active(1).toLeaving()));
+
+    // when - then
+    assertThat(partitionJoinApplier.init(topologyWithPartitionLeaving))
+        .isLeft()
+        .left()
+        .isInstanceOf(IllegalStateException.class);
+
+    // then
+  }
+
+  @Test
+  void shouldRejectJoinIfMemberIsNotActive() {
+    // given
+    final ClusterTopology topologyWithMemberNotActive =
+        initialClusterTopology.updateMember(localMemberId, MemberState::toLeaving);
+
+    // when - then
+    assertThat(partitionJoinApplier.init(topologyWithMemberNotActive))
+        .isLeft()
+        .left()
+        .isInstanceOf(IllegalStateException.class);
+
+    // then
+  }
+
+  @Test
+  void shouldRejectJoinIfPriorityIsNoSet() {
+    // given
+    final PartitionJoinApplier partitionJoinApplier =
+        new PartitionJoinApplier(1, -1, localMemberId, partitionTopologyChangeExecutor);
+
+    // when - then
+    assertThat(partitionJoinApplier.init(initialClusterTopology))
+        .isLeft()
+        .left()
+        .isInstanceOf(IllegalArgumentException.class);
+  }
+
+  @Test
+  void shouldInitializeStateToJoining() {
+    // when
+    final var updater = partitionJoinApplier.init(initialClusterTopology).get();
+    final var resultingTopology = initialClusterTopology.updateMember(localMemberId, updater);
+
+    // then
+    ClusterTopologyAssert.assertThatClusterTopology(resultingTopology)
+        .hasMemberWithPartitions(1, Set.of(1))
+        .member(localMemberId)
+        .hasPartitionWithState(1, new PartitionState(State.JOINING, 1));
+  }
+
+  @Test
+  void shouldExecuteJoinCallBack() {
```

Comment:
:wrench: Add another test case to verify that `partitionJoinApplier.apply(...)` returns a future that fails when `partitionTopologyChangeExecutor.join(...)` fails.